icommands.c

```diff
@@ -98,6 +104,193 @@ enum CommandResult mutt_parse_icommand(/* const */ char *line, struct Buffer *er
   return rc;
 }
 
+/**
+ * dump_map - Dump a map to a buffer
+ * @param buf       Output buffer
+ * @param menu      Map menu
+ * @param map       Map keymap
+ * @param bind      If true dump binds, otherwise macros
+ */
+static void dump_map(struct Buffer *buf, struct Mapping *menu, struct Keymap *map, bool bind)
+{
+  char key_binding[MAX_SEQ];
+  km_expand_key(key_binding, MAX_SEQ, map);
+
+  if (bind)
+  {
+    const char *fn_name = NULL;
+
+    if (map->op == OP_NULL)
+    {
+      mutt_buffer_add_printf(buf, "bind %s %s noop\n", menu->name, key_binding);
+      return;
+    }
+
+    /* The pager and editor menus don't use the generic map,
+     * however for other menus try generic first. */
+    if ((menu->value != MENU_PAGER) && (menu->value != MENU_EDITOR) &&
+        (menu->value != MENU_GENERIC))
+    {
+      fn_name = mutt_get_func(OpGeneric, map->op);
+    }
+
+    /* if it's one of the menus above or generic doesn't find
+     * the function, try with its own menu. */
+    if (!fn_name)
+    {
+      const struct Binding *bindings = km_get_table(menu->value);
+      if (!bindings)
+        return;
+
+      fn_name = mutt_get_func(bindings, map->op);
+    }
+
+    mutt_buffer_add_printf(buf, "bind %s %s %s\n", menu->name, key_binding, fn_name);
+  }
+  else
+  {
+    struct Buffer *tmp = mutt_buffer_new();
+    escape_string(tmp, map->macro);
+
+    if (map->desc)
+    {
+      mutt_buffer_add_printf(buf, "macro %s %s \"%s\" \"%s\"\n", menu->name,
+                             key_binding, tmp->data, map->desc);
+    }
+    else
+    {
+      mutt_buffer_add_printf(buf, "macro %s %s \"%s\"\n", menu->name,
+                             key_binding, tmp->data);
+    }
+
+    mutt_buffer_free(&tmp);
+  }
+}
+
+/**
+ * dump_menu - Dumps all the binds or macros maps of a menu into a buffer
+ * @param buf        Output buffer
+ * @param menu       Menu to dump
+ * @param bind       If true it's :bind, else :macro
+ * @retval bool      true if menu is empty, false if not
+ */
+static bool dump_menu(struct Buffer *buf, struct Mapping *menu, bool bind)
+{
+  bool empty = true;
+  struct Keymap *map = NULL, *next = NULL;
+
+  for (map = Keymaps[menu->value]; map; map = next)
+  {
+    next = map->next;
+
+    if (bind && map->op != OP_MACRO)
+    {
+      empty = false;
+      dump_map(buf, menu, map, true);
+    }
+    else if (!bind && map->op == OP_MACRO)
+    {
+      empty = false;
+      dump_map(buf, menu, map, false);
+    }
+  }
+
+  return empty;
```

Comment:
I like this, but trying out I found out that there is a problem since the initial size of the buffer is 4KiB.

So while the buffer isn't reallocated for the first time, the return always translates to `return (4096 - 4096) == 0`

Reply:
Ah, that makes sense.